test/functional/test_framework/test_node.py

```diff
@@ -151,6 +160,45 @@ def node_encrypt_wallet(self, passphrase):
         self.encryptwallet(passphrase)
         self.wait_until_stopped()
 
+    # def add_p2p_connection(self, p2p_conn_type, dstaddr='127.0.0.1', dstport=None, services=NODE_NETWORK, send_version=True):
+    def add_p2p_connection(self, p2p_conn_type, **kwargs):
+        """Add a p2p connection to the node.
+
+        This method adds the p2p connection to the self.p2ps list and also
+        returns the connection to the caller."""
+        if 'dstport' not in kwargs:
+            kwargs['dstport'] = p2p_port(self.index)
+        if 'dstaddr' not in kwargs:
+            kwargs['dstaddr'] = '127.0.0.1'
+        p2p_conn = p2p_conn_type()
+        self.p2ps.append(p2p_conn)
+        kwargs.update({'rpc': self.rpc, 'callback': p2p_conn})
+        p2p_conn.add_connection(NodeConn(**kwargs))
+
+        return p2p_conn
+
+    @property
+    def p2p(self):
+        """Return the first p2p connection
+
+        Convenience property - most tests only use a single p2p connection to each
+        node, so this saves having to write node.p2ps[0] many times."""
+        assert self.p2ps, "No p2p connection"
+        return self.p2ps[0]
+
+    def send_p2p_message(self, message):
+        """Send a p2p message to the node."""
+        assert self.p2ps, "No p2p connection"
+        self.p2ps[0].send_message(message)
+
+    def disconnect_p2p(self, index=0):
+        """Close the p2p connection to the node."""
+        # Connection could have already been closed by other end. Calling disconnect_p2p()
+        # on an already disconnected p2p connection is not an error.
+        if self.p2ps[index].connection is not None:
```

Comment:
In commit "[tests] Add p2p connection to TestNode"

This is an internal implementation detail of NodeConnCB and you should add a NodeConnCB.disconnect() method to take care of it. This disconnect_p2p method can then be removed and tests call `node.p2p.disconnect()` instead of `node.disconnect_p2p()`.

If for some tests need to clear the `p2ps` array (I don't see any that do) they should just do it directly. This would be nicer than the way the two current disconnect_p2p callers are effectively doing `del p2ps[0]` in a loop to clear the array, for no apparent reason.

Reply:
I like this as it is. TestNode is responsible for appending new `NodeConnCB`s to the `self.p2ps` array. I think it should also be responsible for removing them.